File: Algorithm/Solution/2021-01-15/송원석/Main_1654_랜선자르기.java
package baek;

import java.io.*;
import java.util.*;

public class Main_1654_랜선자르기 {
	
	public static long binarySearch(long left, long right, int[] array, int target) {
		long result = 0;
		
		while(left <= right) {
			long mid = (left+right)/2;
			int sum = 0;
			for(int num : array) {
				sum += num/mid;
			}
			
			if(sum >= target) {
				result = mid;
				left = mid+1;
			}else {
				right = mid-1;
			}
		}
		
		return result;
	}

	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		int K = Integer.parseInt(st.nextToken());
		int N = Integer.parseInt(st.nextToken());
		
		int[] wires = new int[K];
		long total = 0;
		
		for(int i=0;i<K;i++) {
			wires[i] = Integer.parseInt(br.readLine());
			total += wires[i];
		}
		
		long answer = binarySearch(1,total/N,wires,N);
		
		System.out.println(answer);
		
	}

}

File: Algorithm/Solution/2021-01-15/송원석/Main_1654_랜선자르기.js
const fs = require('fs');
// const inputs = fs.readFileSync('./dev/stdin').toString().split("\n");
const inputs = fs.readFileSync('./BOJ/input/input.txt').toString().split('\r\n');

const Main = () => {
  const line = inputs[0].split(" ");
  const K = parseInt(line[0]);
  const N = parseInt(line[1]);
  const wireArray = [];

  let total = 0;

  for (let i = 1; i <= K; i++) {
    const wire = parseInt(inputs[i]);
    wireArray.push(wire);
    total += wire;
  }

  let min = 1;
  let max = Math.floor(total / N);
  let answer = 0;

  const binarySearch = (left, right) => {
    while (left <= right) {
      const mid = Math.floor((left + right) / 2);
      const count = wireArray.reduce((acc, el) => acc += Math.floor(el / mid), 0);
      if (count >= N) {
        answer = mid;
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }
  }

  binarySearch(min, max);

  console.log(answer);
}

Main();